nx/guides/advanced/complex_fft.livemd

```diff
@@ -0,0 +1,297 @@
+|> VegaLite.mark(:bar)
+|> VegaLite.encode_field(:x, "frequencies",
+  type: :quantitative,
+  title: "frequency (Hz)",
+  scale: [domain: [0, 50]]
+)
+|> VegaLite.encode_field(:y, "amplitudes",
+  type: :quantitative,
+  title: "amplitutde",
+  scale: [domain: [0, 30]]
+)
+```
+
+We see the peaks at 5Hz, 20Hz with their amplitudes (the second is half the first).
+
+This is indeed our incomming signal 🎉
+
+We can confirm this visual inspection with a peek into our data. We use `Nx.top_k` function.
+
+```elixir
+{values, indices} = Nx.top_k(amplitudes, k: 5)
+
+filtered_top_frequencies = fn values, indices, threshold ->
+  v = Nx.to_list(values)
+  v0 = Enum.at(v,0)
+  Enum.zip(v, Nx.to_list(indices))
+  |> Enum.filter(fn {v,_idx} -> v >=  v0 * threshold end)
+  |> Enum.unzip()
+  |> elem(1)
+
+end
+
+IO.puts("The top frequencies are: #{inspect(filtered_top_frequencies.(values, indices, 0.001 ))}")
+```
+
+### Visualize the original signal and the IFFT reconstructed
+
+Let's visualize our incoming signal over 400ms. This correspond to 2 periods of our 5Hz signal.
+
+We compute 200 points to have a smooth curve, thus every (400/200=) 2ms.
+
+We also add the "reconstructed" signal via the **Inverse Fast Fourrier Transform** available as `Nx.ifft`.
+
+This gives us 50 values  spaced by (1000 / 50) = 20ms (we sampled as 50Hz during 1s).
+
+We display them as a bar chart next to the "smooth" real signal.
+
+```elixir
+#----------- REAL SIGNAL
+# compute 200 points of the "real" signal during 2/5=400ms (twice the main period)
+
+t = NxSignal.fft_frequencies(2/5, fft_length: 200)
+sample = Signal.source(t)
+
+#----------- RECONSTRUCTED IFFT
+# compute the reconstructed IFFT signal: we get fs=50 points in the range [0..duration=1s]
+y_ifft = Nx.ifft(dft) |> Nx.real()
+l = Nx.size(y_ifft)
+
+# for the graph, we reconstruct  a tensor of 400 points whose values are all zeros
+# except the 400/200 = 20 points spaced every 10th index
+# [y_ifft[0], 0,....,y_fft[1], 0,...., ]
+
+# because the time interval of real signal is 1000/50 = 20ms,
+# and here we draw a point every 400/200= 2ms
+
+
+nb = 20
+
+y_reconstructed =
+  Nx.indexed_put(
+    Nx.broadcast(0.0, {400}),
+    Nx.iota({nb}) |> Nx.dot(10) |> Nx.reshape({nb, 1}),
+    y_ifft[0..nb-1]
+  )
+#----------------
+
+
+data = %{
+  x: Nx.to_list(t),
+  y: Nx.to_list(sample),
+  yr: Nx.to_list(y_reconstructed)
+}
+
+
+VegaLite.new(width: 600, height: 300)
+|> VegaLite.data_from_values(data)
+|> VegaLite.layers([
+  VegaLite.new()
+  |> VegaLite.mark(:line)
+  |> VegaLite.encode_field(:x, "x", type: :quantitative, title: "time (ms)", scale: [domain: [0, 0.4]])
+  |> VegaLite.encode_field(:y, "y", type: :quantitative, title: "signal"),
+  VegaLite.new()
+  |> VegaLite.mark(:bar)
+  |> VegaLite.encode_field(:x, "x", type: :quantitative, scale: [domain: [0, 0.4]])
+  |> VegaLite.encode_field(:y, "yr", type: :quantitative, title: "reconstructed"),
+])
+|> VegaLite.resolve(:scale, y: :independent)
+
```

Comment:
```suggestion
#----------- REAL SIGNAL
# compute 200 points of the "real" signal during 2/5=400ms (twice the main period)

t = Nx.linspace(0, 0.4, n: trunc(0.4 * 500))
sample = Signal.source(t)

#----------- RECONSTRUCTED IFFT
yr = Nx.ifft(dft) |> Nx.real()
fs = 50
tr = Nx.iota(yr.shape) |> Nx.divide(fs)

idx = Nx.less_equal(tr, 0.4)
xr = Nx.select(idx, tr, :nan)
yr = Nx.select(idx, yr, :nan)
#----------------


data = %{
  x: Nx.to_list(t),
  y: Nx.to_list(sample)
}

data_r = %{
  yr: Nx.to_list(yr),
  xr: Nx.to_list(xr)
}

VegaLite.new(width: 600, height: 300)
|> VegaLite.layers([
  VegaLite.new()
  |> VegaLite.data_from_values(data)
  |> VegaLite.mark(:line, tooltip: true)
  |> VegaLite.encode_field(:x, "x", type: :quantitative, title: "time (ms)", scale: [domain: [0, 0.4]])
  |> VegaLite.encode_field(:y, "y", type: :quantitative, title: "signal")
  |> VegaLite.encode_field(:order, "x"),
  VegaLite.new()
  |> VegaLite.data_from_values(data_r)
  |> VegaLite.mark(:bar, tooltip: true)
  |> VegaLite.encode_field(:x, "xr", type: :quantitative, scale: [domain: [0, 0.4]])
  |> VegaLite.encode_field(:y, "yr", type: :quantitative, title: "reconstructed")
  |> VegaLite.encode_field(:order, "xr")
])
```

Reply:
In this approach you plot 2 separate sets of data onto the same graph. I believe this is a cleaner approach that shows that the math itself ensures things coincide.

Reply:
Strange, I tried to superpose 2 sets with different X's, didn't seem to work, but yes in fact.

Now, just stepping back and expressing my POV, the current primitives are no good. I mean, no one has time to spend with 15 LOC. Probably `Tucan` should be used or improved, but you should not need more most 4 LOC with this functional aproach.

Reply:
VegaLite is meant to be lower level and be a 1-1 mapping wuth the JS linrary.

My suggestion works because each layer is based on a separate set of values. I couldn't get it to work with everything in the same layer, but there is a way to do it I think.

You can refactor to use Tucan instead, as it's meant to have a higher level of abstraction

Reply:
Never used Tucan, lets see.